src/libopensc/card-iasecc.c

```diff
@@ -562,65 +557,26 @@ iasecc_init_amos(struct sc_card *card)
 	card->caps |= SC_CARD_CAP_APDU_EXT;
 	card->caps |= SC_CARD_CAP_USE_FCI_AC;
 
+	if (card->type == SC_CARD_TYPE_IASECC_MI)   {
+		rv = iasecc_mi_match(card);
+		if (rv)
+			card->type = SC_CARD_TYPE_IASECC_MI2;
+		else
+			LOG_FUNC_RETURN(ctx, SC_SUCCESS);
```

Comment:
If `iasecc_mi_match()` doesn't succeed, you should not return with `SC_SUCCESS`. I think would break all other cards that are initialized with `iasecc_init_amos()`, right?